indices/download_article_metadata.py

```diff
@@ -1,48 +1,221 @@
+            )
+            if id_subset_len >= retmin * 2:
+                mid = len(id_subset) // 2
+                idq.appendleft(id_subset[:mid])
+                idq.appendleft(id_subset[mid:])
+            else:
+                idq.appendleft(id_subset)
+            time.sleep(error_sleep * successive_errors)
+            continue
 
+        # Write XML to file
+        if initialize_xml:
+            initialize_xml = False
+            write_file.write(f"<{tree.tag}>\n")
+        for child in tree:
+            xml_str = ET.tostring(child, encoding="unicode")
+            write_file.write(xml_str.rstrip() + "\n")
+
+        # Report progress
+        progress_bar.update(id_subset_len)
+
+    progress_bar.close()
+    # Write final line of XML
+    write_file.write(f"</{tree.tag}>\n")
+
+
+def worker(mesh_term, api_key):
     # Make the term safe for putting in a path
-    path_term = args.mesh_term.replace(' ', '_')
+    path_term = mesh_term.replace(' ', '_')
+    path_term = path_term.replace('-', '_')
+    path_term = path_term.replace(',', '')
+    path_term = path_term.lower()
     path = pathlib.Path(f'data/pubmed/efetch/{path_term}.xml.xz')
-
     if not path.exists():
         payload = {
             'db': 'pubmed',
-            'term': f'"journal article"[pt] AND "{args.mesh_term}"[MeSH Terms] AND English[Language]',
+            'term': f'"journal article"[pt] AND "{mesh_term}"[MeSH Terms] AND English[Language]',
         }
-
-        pubmed_ids = esearch_query(payload)
+        pubmed_ids = limited_esearch_query(payload, api_key=api_key, sleep=1, retmax=100000)
         pubmed_ids = sorted(map(int, pubmed_ids))
-        print(f'{len(pubmed_ids):,} articles for {args.mesh_term}')
+        print(f'{len(pubmed_ids):,} articles for {mesh_term}')
 
         path.parent.mkdir(parents=True, exist_ok=True)
         with lzma.open(path, 'wt') as write_file:
             download_pubmed_ids(
                 pubmed_ids, write_file, endpoint='efetch',
-                retmax=200, retmin=50, sleep=0, error_sleep=1,
+                retmax=200, retmin=50, sleep=1, error_sleep=1,
+                api_key=api_key
             )
     else:
-        print(f'{path} exists already; skipping')
\ No newline at end of file
+        print(f'{path} exists already; skipping')
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--key_file', help='A file containing an NCBI API key', default=None)
+    args = parser.parse_args()
+
+    api_key = None
+    if args.key_file is not None:
+        with open(args.key_file, 'r') as in_file:
+            api_key = in_file.readline().strip()
+
+     # TODO read heaidngs from a file
+
+    # I could be a little more memory eficient by creating a partial function or something,
+    # but this is way more readable and I'm only using ~1kB of mem to do this
+    api_keys = [api_key] * len(HEADINGS)
+
+    with ThreadPoolExecutor(max_workers=9) as executor:
```

Comment:
Do you want to pass in `max_workers` as a command line argument? Or will it always be 9?

Reply:
Nope, it'll always be 9 because that's the number that stays below the Entrez ratelimit with our given settings. Funnily enough, with how the global interpreter lock works in Python 9 workers is probably just as much faster on a single-core machine as on an n-core machine